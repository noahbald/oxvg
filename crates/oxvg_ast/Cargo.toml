[package]
name = "oxvg_ast"
version = "0.0.1-beta.1"
edition = "2021"
license.workspace = true
repository.workspace = true
authors.workspace = true
description = "Trait representation for concrete ASTs"

[lints]
workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[features]
default = ["markup5ever", "roxmltree", "visitor"]
markup5ever = ["dep:rcdom", "dep:xml5ever"]
roxmltree = ["dep:roxmltree"]
serialize = []
selectors = ["dep:selectors", "dep:cssparser", "dep:precomputed-hash"]
visitor = ["dep:bitflags"]
style = ["dep:lightningcss", "dep:cssparser_lightningcss", "dep:smallvec"]

[dependencies]
anyhow = { workspace = true, optional = true }
bitflags = { workspace = true, optional = true }
cfg-if = { workspace = true }
cssparser = { workspace = true, optional = true }
cssparser_lightningcss = { package = "cssparser", version = "0.33", optional = true }
derive_more = { workspace = true, features = ["debug"] }
itertools = { workspace = true }
lightningcss = { workspace = true, optional = true }
log = { workspace = true }
markup5ever = { workspace = true }
precomputed-hash = { workspace = true, optional = true }
rcdom = { workspace = true, optional = true }
roxmltree = { workspace = true, optional = true }
# ensure track_alloc_size for wasm compat
servo_arc = { version = "0.4", features = ["track_alloc_size"] }
selectors = { workspace = true, optional = true }
smallvec = { version = "1.13.2", optional = true }
string_cache = { workspace = true }
tendril = { workspace = true }
typed-arena = { workspace = true }
xml5ever = { workspace = true, optional = true }
