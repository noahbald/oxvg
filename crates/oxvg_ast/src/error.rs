//! Error types.

use std::fmt::Display;

/// A parser error
#[derive(Debug)]
pub enum ParseErrorKind<'i> {
    /// Lightningcss failed to parse
    CSSParserError(lightningcss::error::ParserError<'i>),
    /// A fundamental parsing step failed
    Basic(cssparser_lightningcss::BasicParseError<'i>),
    /// A clock value didn't fit the expected range
    InvalidClockValue,
    /// A unexpected set of paint steps were given
    InvalidPaintOrder,
    /// A begin-end syncbase value is missing an id
    MissingSyncbaseId,
}

/// Parse errors that can be encoutered by parsing
pub type ParseError<'input> = cssparser_lightningcss::ParseError<'input, ParseErrorKind<'input>>;

impl<'input> ParseErrorKind<'input> {
    /// Maps a lightnincss error to `Self`
    pub fn from_css(
        error: cssparser_lightningcss::ParseError<'input, lightningcss::error::ParserError<'input>>,
    ) -> cssparser_lightningcss::ParseError<'input, ParseErrorKind<'input>> {
        match error.kind {
            cssparser_lightningcss::ParseErrorKind::Basic(e) => {
                cssparser_lightningcss::ParseError {
                    kind: cssparser_lightningcss::ParseErrorKind::Basic(e),
                    location: error.location,
                }
            }
            cssparser_lightningcss::ParseErrorKind::Custom(e) => {
                cssparser_lightningcss::ParseError {
                    kind: cssparser_lightningcss::ParseErrorKind::Custom(
                        ParseErrorKind::CSSParserError(e),
                    ),
                    location: error.location,
                }
            }
        }
    }

    /// Maps a basic error to `Self`
    pub fn from_basic(
        error: cssparser_lightningcss::BasicParseError<'input>,
    ) -> cssparser_lightningcss::ParseError<'input, Self> {
        cssparser_lightningcss::ParseError {
            kind: cssparser_lightningcss::ParseErrorKind::Basic(error.kind),
            location: error.location,
        }
    }

    /// Maps a parser error to `Self`
    pub fn from_parser(
        error: cssparser_lightningcss::ParseError<'input, cssparser_lightningcss::BasicParseError>,
    ) -> cssparser_lightningcss::ParseError<'input, Self> {
        match error.kind {
            cssparser_lightningcss::ParseErrorKind::Basic(e) => {
                cssparser_lightningcss::ParseError {
                    kind: cssparser_lightningcss::ParseErrorKind::Basic(e),
                    location: error.location,
                }
            }
            cssparser_lightningcss::ParseErrorKind::Custom(e) => {
                cssparser_lightningcss::ParseError {
                    kind: cssparser_lightningcss::ParseErrorKind::Custom(ParseErrorKind::Basic(e)),
                    location: error.location,
                }
            }
        }
    }
}

/// A printer error
pub type PrinterError = lightningcss::error::PrinterError;

/// An error while serializing a document.
#[derive(Debug)]
pub enum XmlWriterError {
    /// An error while running an io operation.
    IO(std::io::Error),
    /// An error while formatting
    FMT(std::fmt::Error),
    /// An error while flushing buffer.
    BufWriter(std::io::IntoInnerError<std::io::BufWriter<Vec<u8>>>),
    /// An error after writing to string.
    UTF8(std::string::FromUtf8Error),
    /// An error while serializing an attribute
    PrinterError(PrinterError),
    /// Did not have opening element name when closing element.
    ClosedUnopenedElement,
    /// Attempted to write attribute before `start_element()` or after `close_element()`.
    AttributeWrittenBeforeElement,
    /// Declaration was already written.
    DeclarationAlreadyWritten,
    /// Attempts to write text before `start_element()`.
    TextBeforeElement,
    /// Attempts to write CDATA with `]]>` in the content.
    BadCDATA,
}
impl std::fmt::Display for XmlWriterError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::IO(err) => err.fmt(f),
            Self::FMT(err) => err.fmt(f),
            Self::BufWriter(err) => err.fmt(f),
            Self::UTF8(err) => err.fmt(f),
            Self::PrinterError(err) => err.fmt(f),
            Self::ClosedUnopenedElement => {
                "Did not have opening element name when closing element.".fmt(f)
            }
            Self::AttributeWrittenBeforeElement => {
                "Attempted to write attribute before `start_element()` or after `close_element()`."
                    .fmt(f)
            }
            Self::TextBeforeElement => "Attempts to write text before `start_element()`.".fmt(f),
            Self::BadCDATA => "Attempts to write CDATA with `]]>` in the content.".fmt(f),
            Self::DeclarationAlreadyWritten => "Declaration was already written.".fmt(f),
        }
    }
}
impl std::error::Error for XmlWriterError {}

/// An error while gathering computed styles
#[derive(Debug, Clone)]
pub enum ComputedStylesError<'input> {
    /// A selector was found that may affect validity of computed styles
    BadSelector(lightningcss::selector::SelectorList<'input>),
}
impl Display for ComputedStylesError<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::BadSelector(selector) => {
                f.write_str("Found an invalid selector while processing: ")?;
                selector.fmt(f)
            }
        }
    }
}
impl std::error::Error for ComputedStylesError<'_> {}
