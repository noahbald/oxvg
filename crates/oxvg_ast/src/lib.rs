#[macro_use]
extern crate bitflags;

#[cfg(feature = "style")]
#[macro_use]
extern crate smallvec;

pub mod atom;
pub mod attribute;
pub mod class_list;
pub mod document;
pub mod element;
pub mod implementations;
pub mod name;
pub mod node;

#[cfg(feature = "visitor")]
pub mod visitor;

#[cfg(feature = "parse")]
pub mod parse;

#[cfg(feature = "serialize")]
pub mod serialize;

#[cfg(feature = "selectors")]
pub mod selectors;

#[cfg(feature = "style")]
pub mod style;
