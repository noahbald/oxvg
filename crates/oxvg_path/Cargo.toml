[package]
name = "oxvg_path"
version = "0.0.1-beta.4"
edition = "2021"
license.workspace = true
repository.workspace = true
authors.workspace = true
readme = "../../readme.md"
keywords = ["svg", "svgo", "path"]
description = "SVG path parsing, serialization, and minification"

[lints]
workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[features]
default = ["parse", "format", "optimise"]
jsonschema = ["schemars", "serde"]
oxvg = ["dep:lightningcss", "dep:oxvg_style"]
serde = ["dep:serde"]
parse = ["bitflags"]
format = ["ryu"]
optimise = ["format", "bitflags"]

[dependencies]
oxvg_style = { workspace = true, optional = true }
bitflags = { version = "2.6.0", optional = true }
lightningcss = { workspace = true, optional = true }
serde = { workspace = true, optional = true }
schemars = { version = "0.8.19", optional = true }
ryu = { version = "1.0.18", optional = true }

[dev-dependencies]
insta = { workspace = true }
