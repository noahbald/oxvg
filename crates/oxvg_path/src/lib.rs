//! OXVG Path is a library used for parsing and minifying SVG paths.
//! It supports parsing of any valid SVG path and provides optimisations close to exactly as SVGO.
//!
//! Use the [Path](Path) struct for simple parsing and serializing. By only parsing and serializing,
//! it will produce optimised formatting out of the box.
//! It is made up individual command [Data](command::Data).
//!
//! For more rigorous minification, try using the [run](convert::run) function. This will use
//! non-destructive conversions to shorten the path.
//!
//! # Differences to SVGO
//!
//! - Unlike SVGO, all close paths are serialized as `Z` instead of either `z` or `Z`. This is fine because the two commands function exactly the same.
//! - An equivalent of the `applyTransforms` option isn't available, but may be in the future.
//!
//! # Licensing
//!
//! This library is based off the [`convertPathData`](https://svgo.dev/docs/plugins/convertPathData/) plugin from SVGO and is similarly released under MIT.
#[cfg(feature = "optimise")]
#[cfg(feature = "parse")]
#[macro_use]
extern crate bitflags;

#[cfg_attr(feature = "napi", macro_use)]
extern crate napi_derive;

#[cfg(feature = "optimise")]
pub mod command;
#[cfg(feature = "optimise")]
pub mod convert;
#[cfg(feature = "optimise")]
pub mod geometry;
#[cfg(feature = "optimise")]
pub(crate) mod math;
#[cfg(feature = "parse")]
pub mod parser;
#[cfg(feature = "optimise")]
pub mod points;
#[cfg(feature = "optimise")]
pub mod positioned;

use points::{Point, Points};

#[cfg(feature = "parse")]
use crate::parser::Parser;

#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "jsonschema", derive(schemars::JsonSchema))]

/// A path is a set of commands
///
/// # Example
///
/// Out of the box, parsing and serializing a path will produce optimal formatting
///
/// ```
/// use oxvg_path::Path;
///
/// let path = Path::parse("M 10 0.01 L 0.5 -1").unwrap();
/// assert_eq!(&path.to_string(), "M10 .01.5-1");
/// ```
///
/// For more extensive minification, look into using the [run](convert::run) function.
pub struct Path(pub Vec<command::Data>);

impl Path {
    #[cfg(feature = "parse")]
    /// Parses a path definition from a string
    ///
    /// # Errors
    /// If the definition is invalid
    pub fn parse(definition: &str) -> Result<Self, parser::Error> {
        Parser::default().parse(definition)
    }

    /// Checks if two paths have an intersection by checking convex hulls collision using
    /// Gilbert-Johnson-Keerthi distance algorithm.
    ///
    /// # Panics
    /// If internal assertions fail
    pub fn intersects(&self, other: &Self) -> bool {
        let points_1 = Points::from_positioned(&convert::relative(self));
        let points_2 = Points::from_positioned(&convert::relative(other));

        // First check whether their bounding box intersects
        if points_1.max_x <= points_2.min_x
            || points_2.max_x <= points_1.min_x
            || points_1.max_y <= points_2.min_y
            || points_2.max_y <= points_1.min_y
            || points_1.list.iter().all(|set_1| {
                points_2.list.iter().all(|set_2| {
                    set_1.list[set_1.max_x].0[0] <= set_2.list[set_2.min_x].0[0]
                        || set_2.list[set_2.max_x].0[0] <= set_1.list[set_1.min_x].0[0]
                        || set_1.list[set_1.max_y].0[1] <= set_2.list[set_2.min_y].0[1]
                        || set_2.list[set_2.max_y].0[1] <= set_1.list[set_1.min_y].0[1]
                })
            })
        {
            log::debug!("no intersection, bounds check failed");
            return false;
        }

        // i.e. https://en.wikipedia.org/wiki/Gilbert%E2%80%93Johnson%E2%80%93Keerthi_distance_algorithm
        let mut hull_nest_1 = points_1.list.iter().map(Point::convex_hull);
        let hull_nest_2: Vec<_> = points_2.list.iter().map(Point::convex_hull).collect();

        hull_nest_1.any(|hull_1| {
            if hull_1.list.len() < 3 {
                return false;
            }

            hull_nest_2.iter().any(|hull_2| {
                if hull_2.list.len() < 3 {
                    return false;
                }

                let mut simplex = vec![hull_1.get_support(hull_2, geometry::Point([1.0, 0.0]))];
                let mut direction = simplex[0].minus();
                let mut iterations = 10_000;

                loop {
                    iterations -= 1;
                    if iterations == 0 {
                        log::error!("Infinite loop while finding path intersections");
                        return true;
                    }
                    simplex.push(hull_1.get_support(hull_2, direction));
                    if direction.dot(simplex.last().unwrap()) <= 0.0 {
                        return false;
                    }
                    if geometry::Point::process_simplex(&mut simplex, &mut direction) {
                        return true;
                    }
                }
            })
        })
    }
}

#[cfg(feature = "format")]
impl std::fmt::Display for Path {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use std::fmt::Write;

        if self.0.len() == 1 {
            self.0.first().unwrap().fmt(f)?;
            return Ok(());
        }
        self.0
            .windows(2)
            .enumerate()
            .try_for_each(|(i, window)| -> std::fmt::Result {
                let prev = &window[0];
                let current = &window[1];
                if i == 0 {
                    prev.fmt(f)?;
                }
                let str = current.to_string();
                if current.is_space_needed(prev) && !str.starts_with('-') {
                    f.write_char(' ')?;
                }
                f.write_str(&str)?;
                Ok(())
            })
    }
}

#[cfg(feature = "format")]
impl From<Path> for String {
    fn from(value: Path) -> Self {
        format!("{value}")
    }
}

#[cfg(feature = "format")]
impl From<&Path> for String {
    fn from(value: &Path) -> Self {
        format!("{value}")
    }
}

#[test]
#[cfg(feature = "default")]
fn test_path_parse() {
    // Should parse single command
    insta::assert_snapshot!(Path::parse("M 10,50").unwrap());

    // Should parse multiple commands
    insta::assert_snapshot!(
        Path::parse("M 10,50 C 20,30 40,50 60,70 C 10,20 30,40 50,60").unwrap()
    );

    // Should parse arc
    insta::assert_snapshot!(Path::parse("m-0,1a 25,25 -30 0,1 0,0").unwrap());

    // Should parse implicit
    insta::assert_snapshot!(
        Path::parse("M 10,50 C 1,2 3,4 5,6.5 .1 .2 .3 .4 .5 -.05176e-005").unwrap()
    );

    // Should parse minified
    insta::assert_snapshot!(Path::parse("M10 50C1 2 3 4 5 6.5.1.2.3.4.5-5.176e-7").unwrap());

    // Should error when command isn't given
    assert!(Path::parse("0,0").is_err());

    // Should error when args are missing
    assert!(Path::parse("m1").is_err());
}
