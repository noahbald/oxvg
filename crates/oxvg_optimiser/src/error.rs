use std::{fmt::Display, sync::mpsc::RecvTimeoutError};

use oxvg_ast::{atom::Atom, error::ComputedStylesError};

#[derive(Debug, Clone)]
pub enum PrecheckError<'input> {
    ScriptingNotSupported,
    AnimationNotSupported,
    ConditionalProcessingNotSupported,
    ReferencesExternalXLink(Atom<'input>),
}
impl Display for PrecheckError<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ScriptingNotSupported => f.write_str("scripting is not supported"),
            Self::AnimationNotSupported => f.write_str("animation is not supported"),
            Self::ConditionalProcessingNotSupported => {
                f.write_str("conditional processing attributes is not supported")
            }
            Self::ReferencesExternalXLink(xlink) => f.write_fmt(format_args!("the `xlink:href` attribute is referencing an external object '{xlink}' which is not supported"))
        }
    }
}
impl std::error::Error for PrecheckError<'_> {}

#[derive(Debug, Clone)]
#[cfg(feature = "napi")]
pub enum PrefixGeneratorError {
    ThreadTimeoutError(RecvTimeoutError),
    ThreadUnknownError,
    NapiError(napi::Status),
    ClientError(String),
}

#[derive(Debug, Clone)]
pub enum JobsError<'input> {
    ComputedStylesError(ComputedStylesError<'input>),
    CleanupValuesPrecision(u8),
    Precheck(PrecheckError<'input>),
    #[cfg(feature = "napi")]
    PrefixGenerator(PrefixGeneratorError),
    InvalidUserSelector(String),
    InvalidUserRegex(regex::Error),
    #[deprecated]
    Generic(String),
}
impl Display for JobsError<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ComputedStylesError(e) => e.fmt(f),
            Self::CleanupValuesPrecision(n) => f.write_fmt(format_args!(
                "The float-precision `{n}` is larger than the maximum of 5"
            )),
            Self::Precheck(e) => e.fmt(f),
            #[cfg(feature = "napi")]
            Self::PrefixGenerator(_) => f.write_str("Prefix generate failed to close threads"),
            Self::InvalidUserSelector(e) => {
                f.write_fmt(format_args!("Invalid selector in configuration: {e}"))
            }
            Self::InvalidUserRegex(e) => e.fmt(f),
            Self::Generic(e) => e.fmt(f),
        }
    }
}
impl std::error::Error for JobsError<'_> {}
