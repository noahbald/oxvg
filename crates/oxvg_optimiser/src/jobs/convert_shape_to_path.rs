use std::cell;

use itertools::peek_nth;
use lightningcss::values::length::{Length, LengthOrNumber};
use oxvg_ast::{
    attribute::data::{path, presentation::LengthPercentage, uncategorised::Radius, AttrId},
    element::{data::ElementId, Element},
    get_attribute, has_attribute, set_attribute,
    visitor::{Context, Info, Visitor},
};
use oxvg_path::{command::Data, convert, Path};
use serde::{Deserialize, Serialize};

use crate::error::JobsError;

use super::convert_path_data::ConvertPrecision;

#[cfg(feature = "wasm")]
use tsify::Tsify;

#[cfg_attr(feature = "wasm", derive(Tsify))]
#[cfg_attr(feature = "napi", napi(object))]
#[derive(Deserialize, Serialize, Debug, Default, Clone)]
#[serde(rename_all = "camelCase")]
/// Converts basic shapes to `<path>` elements
///
/// # Correctness
///
/// Rounding errors may cause slight changes in visual appearance.
///
/// # Errors
///
/// Never.
///
/// If this job produces an error or panic, please raise an [issue](https://github.com/noahbald/oxvg/issues)
pub struct ConvertShapeToPath {
    /// Whether to convert `<circle>` and `<ellipses>` to paths.
    #[serde(default = "default_convert_arcs")]
    pub convert_arcs: bool,
    /// The number of decimal places to round to
    #[cfg_attr(feature = "wasm", tsify(type = "null | false | number"))]
    #[serde(default = "ConvertPrecision::default")]
    pub float_precision: ConvertPrecision,
}

impl<'input, 'arena> Visitor<'input, 'arena> for ConvertShapeToPath {
    type Error = JobsError<'input>;

    fn element(
        &self,
        element: &Element<'input, 'arena>,
        context: &mut Context<'input, 'arena, '_>,
    ) -> Result<(), Self::Error> {
        let name = element.qual_name();

        let path_options = &convert::Options {
            precision: self.float_precision.0,
            ..convert::Options::default()
        };
        let convert_arcs = self.convert_arcs;

        match name {
            ElementId::Rect => Self::rect_to_path(element, path_options, context.info),
            ElementId::Line => Self::line_to_path(element, path_options, context.info),
            ElementId::Polyline => Self::poly_to_path(element, path_options, false, context.info),
            ElementId::Polygon => Self::poly_to_path(element, path_options, true, context.info),
            ElementId::Circle if convert_arcs => {
                Self::circle_to_path(element, path_options, context.info)
            }
            ElementId::Ellipse if convert_arcs => {
                Self::ellipse_to_path(element, path_options, context.info)
            }
            _ => {}
        }
        Ok(())
    }
}

fn lp_px(lp: cell::Ref<LengthPercentage>) -> Option<f64> {
    use lightningcss::values::length::LengthPercentage;
    match &lp.0 {
        LengthPercentage::Dimension(d) => d.to_px().map(|px| px as f64),
        _ => None,
    }
}
fn ln_px(ln: cell::Ref<LengthOrNumber>) -> Option<f64> {
    match &*ln {
        LengthOrNumber::Number(n) => Some(*n as f64),
        LengthOrNumber::Length(Length::Value(l)) => l.to_px().map(|px| px as f64),
        LengthOrNumber::Length(Length::Calc(_)) => None,
    }
}
fn r_px(r: cell::Ref<Radius>) -> Option<f64> {
    cell::Ref::filter_map(r, |r| match r {
        Radius::LengthPercentage(lp) => Some(lp),
        Radius::Auto => None,
    })
    .ok()
    .and_then(lp_px)
}
impl ConvertShapeToPath {
    fn rect_to_path<'input, 'arena>(
        element: &Element<'input, 'arena>,
        options: &convert::Options,
        info: &Info<'input, 'arena>,
    ) {
        if has_attribute!(element, RX | RY) {
            return;
        }

        let Some(x) = get_attribute!(element, X).and_then(lp_px) else {
            return;
        };
        let Some(y) = get_attribute!(element, Y).and_then(lp_px) else {
            return;
        };
        let Some(width) = get_attribute!(element, Width).and_then(lp_px) else {
            return;
        };
        let Some(height) = get_attribute!(element, Height).and_then(lp_px) else {
            return;
        };

        let mut path = Path(vec![
            Data::MoveTo([x, y]),
            Data::HorizontalLineTo([x + width]),
            Data::VerticalLineTo([y + height]),
            Data::HorizontalLineTo([x]),
            Data::ClosePath,
        ]);
        options.round_path(&mut path, options.error());

        set_attribute!(element, D(path::Path(path)));
        element.remove_attribute(&AttrId::X);
        element.remove_attribute(&AttrId::Y);
        element.remove_attribute(&AttrId::Width);
        element.remove_attribute(&AttrId::Height);
        element.set_local_name(ElementId::Path, &info.allocator);
    }

    fn line_to_path<'input, 'arena>(
        element: &Element<'input, 'arena>,
        options: &convert::Options,
        info: &Info<'input, 'arena>,
    ) {
        let Some(x1) = get_attribute!(element, X1).and_then(ln_px) else {
            return;
        };
        let Some(y1) = get_attribute!(element, Y1).and_then(ln_px) else {
            return;
        };
        let Some(x2) = get_attribute!(element, X2).and_then(ln_px) else {
            return;
        };
        let Some(y2) = get_attribute!(element, Y2).and_then(ln_px) else {
            return;
        };

        let mut path = Path(vec![
            Data::MoveTo([x1, y1]),
            Data::Implicit(Box::new(Data::LineTo([x2, y2]))),
        ]);
        options.round_path(&mut path, options.error());

        set_attribute!(element, D(path::Path(path)));
        element.remove_attribute(&AttrId::X1);
        element.remove_attribute(&AttrId::Y1);
        element.remove_attribute(&AttrId::X2);
        element.remove_attribute(&AttrId::Y2);
        element.set_local_name(ElementId::Path, &info.allocator);
    }

    fn poly_to_path<'input, 'arena>(
        element: &Element<'input, 'arena>,
        options: &convert::Options,
        is_polygon: bool,
        info: &Info<'input, 'arena>,
    ) {
        let Some(coords) = get_attribute!(element, Points) else {
            return;
        };
        let mut coords = peek_nth(coords.list.iter());

        if coords.peek_nth(3).is_none() {
            element.remove();
            return;
        }

        let mut data = Vec::with_capacity(2);
        while let (Some(x), Some(y)) = (coords.next(), coords.next()) {
            let x = *x as f64;
            let y = *y as f64;
            let command = if data.is_empty() {
                Data::MoveTo([x, y])
            } else {
                Data::Implicit(Box::new(Data::LineTo([x, y])))
            };
            data.push(command);
        }
        if is_polygon {
            data.push(Data::ClosePath);
        }
        let mut path = Path(data);
        options.round_path(&mut path, options.error());

        set_attribute!(element, D(path::Path(path)));
        element.remove_attribute(&AttrId::Points);
        element.set_local_name(ElementId::Path, &info.allocator);
    }

    #[allow(clippy::similar_names)]
    fn circle_to_path<'input, 'arena>(
        element: &Element<'input, 'arena>,
        options: &convert::Options,
        info: &Info<'input, 'arena>,
    ) {
        let Some(cx) = get_attribute!(element, CX).and_then(lp_px) else {
            return;
        };
        let Some(cy) = get_attribute!(element, CY).and_then(lp_px) else {
            return;
        };
        let Some(r) = get_attribute!(element, RCircle).and_then(lp_px) else {
            return;
        };

        let mut path = Path(vec![
            Data::MoveTo([cx, cy - r]),
            Data::ArcTo([r, r, 0.0, 1.0, 0.0, cx, cy + r]),
            Data::Implicit(Box::new(Data::ArcTo([r, r, 0.0, 1.0, 0.0, cx, cy - r]))),
            Data::ClosePath,
        ]);
        options.round_path(&mut path, options.error());

        set_attribute!(element, D(path::Path(path)));
        element.remove_attribute(&AttrId::CX);
        element.remove_attribute(&AttrId::CY);
        element.remove_attribute(&AttrId::RCircle);
        element.set_local_name(ElementId::Path, &info.allocator);
    }

    #[allow(clippy::similar_names)]
    fn ellipse_to_path<'input, 'arena>(
        element: &Element<'input, 'arena>,
        options: &convert::Options,
        info: &Info<'input, 'arena>,
    ) {
        let Some(cx) = get_attribute!(element, CX).and_then(lp_px) else {
            return;
        };
        let Some(cy) = get_attribute!(element, CY).and_then(lp_px) else {
            return;
        };
        let Some(rx) = get_attribute!(element, RX).and_then(r_px) else {
            return;
        };
        let Some(ry) = get_attribute!(element, RY).and_then(r_px) else {
            return;
        };

        let mut path = Path(vec![
            Data::MoveTo([cx, cy - ry]),
            Data::ArcTo([rx, ry, 0.0, 1.0, 0.0, cx, cy + ry]),
            Data::Implicit(Box::new(Data::ArcTo([rx, ry, 0.0, 1.0, 0.0, cx, cy - ry]))),
            Data::ClosePath,
        ]);
        options.round_path(&mut path, options.error());

        set_attribute!(element, D(path::Path(path)));
        element.remove_attribute(&AttrId::CX);
        element.remove_attribute(&AttrId::CY);
        element.remove_attribute(&AttrId::RX);
        element.remove_attribute(&AttrId::RY);
        element.set_local_name(ElementId::Path, &info.allocator);
    }
}

const fn default_convert_arcs() -> bool {
    false
}

#[test]
fn convert_shape_to_path() -> anyhow::Result<()> {
    use crate::test_config;

    insta::assert_snapshot!(test_config(
        r#"{ "convertShapeToPath": {} }"#,
        Some(
            r##"<svg xmlns="http://www.w3.org/2000/svg">
    <rect width="100%"/>
    <rect width="100%" height="100%"/>
    <rect x="25%" y="25%" width="50%" height="50%"/>
    <rect x="25pt" y="25pt" width="50pt" height="50pt"/>
    <rect x="10" y="10" width="50" height="50" rx="4"/>
    <rect x="0" y="0" width="20" height="20" ry="5"/>
    <rect width="32" height="32"/>
    <rect x="20" y="10" width="50" height="40"/>
    <rect fill="#666" x="10" y="10" width="10" height="10"/>
</svg>
"##
        ),
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "convertShapeToPath": {} }"#,
        Some(
            r##"<svg xmlns="http://www.w3.org/2000/svg">
    <line x2="100%" y2="100%"/>
    <line x1="24" y2="24"/>
    <line x1="10" y1="10" x2="50" y2="20"/>
    <line stroke="#000" x1="10" y1="10" x2="50" y2="20"/>
</svg>"##
        ),
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "convertShapeToPath": {} }"#,
        Some(
            r##"<svg xmlns="http://www.w3.org/2000/svg">
    <polyline points="10,10 20"/>
    <polyline points="10,80 20,50 50,20 80,10"/>
    <polyline points="20 ,10  50    40 30.5-1e-1 , 20 10"/>
    <polyline stroke="#000" points="10,10 20,20 10,20"/>
    <polygon points="10,10 20"/>
    <polygon points="10,80 20,50 50,20 80,10"/>
    <polygon points="20 10  50 40 30,20"/>
    <polygon stroke="#000" points="10,10 20,20 10,20"/>
    <polygon stroke="none" points="10,10 20,20 10,20"/>
</svg>"##
        ),
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "convertShapeToPath": { "convertArcs": true } }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <circle cx="10" cy="10" r="5"/>
    <ellipse cx="10" cy="10" rx="5" ry="5"/>
</svg>"#
        ),
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "convertShapeToPath": { "convertArcs": true, "floatPrecision": 3 } }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg" width="65mm" height="45mm" viewBox="0 0 65 45">
  <rect x="26.614" y="29.232" width="34.268" height="8.1757"/>
  <line x1="26.6142" y1="29.2322" x2="34.2682" y2="8.1757"/>
  <polyline points="26.6142,29.2322 34.2682,8.1757"/>
  <polygon points="26.6142,29.2322 34.2682,8.1757"/>
  <circle cx="26.6142" cy="29.2322" r="34.2682"/>
  <ellipse cx="26.6142" cy="29.2322" rx="34.2682" ry="8.1757"/>
</svg>"#
        ),
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "convertShapeToPath": {} }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
  <defs>
    <rect id="rect1" width="120" height="120" />
  </defs>
</svg>"#
        ),
    )?);

    Ok(())
}
