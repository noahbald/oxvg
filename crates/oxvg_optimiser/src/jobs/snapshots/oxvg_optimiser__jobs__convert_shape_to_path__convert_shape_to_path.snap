---
source: crates/oxvg_optimiser/src/jobs/convert_shape_to_path.rs
expression: "test_config(r#\"{ \"convertShapeToPath\": {} }\"#,\nSome(r##\"<svg xmlns=\"http://www.w3.org/2000/svg\">\n    <rect width=\"100%\"/>\n    <rect width=\"100%\" height=\"100%\"/>\n    <rect x=\"25%\" y=\"25%\" width=\"50%\" height=\"50%\"/>\n    <rect x=\"25pt\" y=\"25pt\" width=\"50pt\" height=\"50pt\"/>\n    <rect x=\"10\" y=\"10\" width=\"50\" height=\"50\" rx=\"4\"/>\n    <rect x=\"0\" y=\"0\" width=\"20\" height=\"20\" ry=\"5\"/>\n    <rect width=\"32\" height=\"32\"/>\n    <rect x=\"20\" y=\"10\" width=\"50\" height=\"40\"/>\n    <rect fill=\"#666\" x=\"10\" y=\"10\" width=\"10\" height=\"10\"/>\n</svg>\n\"##),)?"
---
<svg xmlns="http://www.w3.org/2000/svg">
    <rect width="100%"/>
    <rect width="100%" height="100%"/>
    <rect x="25%" y="25%" width="50%" height="50%"/>
    <rect x="25pt" y="25pt" width="50pt" height="50pt"/>
    <rect x="10" y="10" width="50" height="50" rx="4"/>
    <rect x="0" y="0" width="20" height="20" ry="5"/>
    <path d="M0 0H32V32H0Z"/>
    <path d="M20 10H70V50H20Z"/>
    <path fill="#666" d="M10 10H20V20H10Z"/>
</svg>
